### PR TITLE
Added Graupner Hott telemetry

### PR DESCRIPTION
**Pull request by tridge**
This adds support for Graupner Hott telemetry. Hott is a UART telem protocol running at 19200 on a half-duplex UART.
To test setup a serial port with:
 - SERIALn_PROTOCOL=26
 - SERIALn_OPTIONS=4

This PR is waiting testing by @lvale

### CODE REVIEW

**Comment by Rolf-G** [comment]
Is a simple cable between AP Serial-Tx and Graupner-RX Sensor pin sufficient or do we need some additional hardeware like this; https://discuss.ardupilot.org/t/some-soldering-required/27613

**Comment by tridge** [comment]
@Rolf-G You don't need any additional hardware, just a simple cable between the receiver sensor pin and the TX pin of a UART on the flight controller. Then you need to set SERIALn_OPTIONS=4 to set the port as half-duplex and set SERIALn_PROTOCOL=27
It is much simpler than FrSky as the protocol is not inverted, which means it works with all STM32 types, and is not affected by the level shifters on some boards.

**Comment by Rolf-G** [comment]
Just tested MX16/GR-12 with F405 Wing with todays ArduPlane V4.1.0dev (a2d40d04)
on serial 3 and 5 with different baud settings - unfortunately without success. Sensors were not recognized. Cross-check with a Graupner Vario connected to the sensor pin was recognized immediately.

**Comment by tridge** [comment]
> Just tested MX16/GR-12 with F405 Wing with todays ArduPlane V4.1.0dev ([a2d40d0](https://github.com/ArduPilot/ardupilot/commit/a2d40d040af9822f46b8716523b9cdd53363da05))
> on serial 3 and 5 with 19200 and also different baud settings - unfortunately without success. 

I'm using the same combination of transmitter and receiver, but I have a smart-box connected to the transmitter. I notice that sometimes the smart box doesn't request the messages that ArduPilot supports.
I'm hoping Ralf from Graupner will test soon and help me work out how to make this more robust.

> Cross-check with a Graupner Vario connected to the sensor pin was recognized immediately.

is there any chance you could capture what the vario sends?

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+        uint8_t stop_byte = 0x7d;
+    } msg {};
+
+    AP_GPS &gps = AP::gps();
+    Location loc;
+
+    {
+        WITH_SEMAPHORE(gps.get_semaphore());
+        loc = gps.location();
+        msg.flight_direction = uint16_t(gps.ground_course() * 0.5 + 0.5);
```
Comment (peterbarker) [inline libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: Should these come from the AHRS instead?

The protocol writers were probably assuming the flight controller would be simply regurgitating GPS information.

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+    msg.pos_NS_dm = dm;
+    msg.pos_NS_sec = sec;
+
+    GPS_to_DMS(lon, msg.pos_EW, dm, sec);
+    msg.pos_EW_dm = dm;
+    msg.pos_EW_sec = sec;
+
+    AP_AHRS &ahrs = AP::ahrs();
+    Vector2f home_vec;
+    float posD = 0;
```
Comment (peterbarker) [inline libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: Slight oddity here in that we set other message fields while still holding the semaphore - why not this one?  Or not set the message fields generally while holding the semaphore.

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        msg.gps_fix_char = '3';
+        break;
+    }
+    msg.gps_fix_char2 = msg.gps_fix_char;
+
+    msg.home_direction = degrees(atan2f(home_vec.y, home_vec.x)) * 0.5 + 0.5;
+
+    AP_RTC &rtc = AP::rtc();
+    {
+        WITH_SEMAPHORE(rtc.get_semaphore());
```
Comment (peterbarker) [inline libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: I wonder if we should make taking the semaphore (recursively) implicit in `get_system_clock_utc`

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+        uint8_t horiz_acc;            //#39 horizontal accuracy dm
+        uint16_t msl_altitude;        //#40 mean sea level altitude
+        uint8_t gps_fix_char2;        //#42 GPS fix character2??. display, 'D' = DGPS, '2' = 2D, '3' = 3D, '-' = no fix
+        uint8_t version;       //#43  0: GPS Graupner #33600
+        uint8_t status1;
+        uint8_t status2;
+        uint8_t stop_byte = 0x7d;
+    } msg {};
+
+    AP_GPS &gps = AP::gps();
```
Comment (peterbarker) [inline libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: You're taking information from this object after releasing the semaphore.

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+}
+
+/*
+  send a packet out
+ */
+void AP_Hott_Telem::send_packet(const uint8_t *b, uint8_t len)
+{
+    uint8_t crc = 0;
+    while (len) {
+        hal.scheduler->delay_microseconds(BYTE_DELAY_US);
```
Comment (peterbarker) [inline libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: Should we be starting a timeout based on when `tx_pending` returns false instead?

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+        uint32_t n = uart->available();
+        if (n < 2) {
+            // wait for 2 bytes
+            continue;
+        }
+        if (n > 2) {
+            while (n--) {
+                uart->read();
+                hal.scheduler->delay_microseconds(20);
```
Comment (peterbarker) [inline libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: This I don't understand.  We need to drain the uart slowly?!

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+#include <AP_Baro/AP_Baro.h>
+#include <AP_RPM/AP_RPM.h>
+#include <AP_Airspeed/AP_Airspeed.h>
+#include <AP_Stats/AP_Stats.h>
+#include <AP_GPS/AP_GPS.h>
+#include <AP_RTC/AP_RTC.h>
+#include <AP_InertialSensor/AP_InertialSensor.h>
+#include <stdio.h>
+
+#define PROT_BINARY   0x80
```
Comment (peterbarker) [inline libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: Would be nice to use an enum class here.

**Comment by Rolf-G** [comment]
Hi Tridge, it works ! (I accidentally didn't compiled your changes, just compiled Master - sorry)

**Comment by tridge** [comment]
> Hi Tridge, it works ! (I accidentally didn't compiled your changes, just compiled Master - sorry)

great!

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: given the use of accuracy info I think this is really intended to be GPS

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: when putting a uint32_t or less it is atomic anyway

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp (hunk 6 above)
Comment (tridge) [reply libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: yes, this thread is high priority. If a bug leads us to get a huge number we don't want to chew all CPU draining 100M bytes

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp (hunk 7 above)
Comment (tridge) [reply libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: there are more possible values and we deliberately don't handle all of them. I don't think an enum class is a win here

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+    if (ahrs.get_velocity_NED(vel)) {
+        msg.climbrate = uint16_t(30000.5 + vel.z * -100);
+        msg.climbrate3s = 120 + vel.z * -3;
+    }
+    AP_RPM *rpm = AP::rpm();
+    if (rpm && rpm->enabled(0)) {
+        msg.rpm = rpm->get_rpm(0) / 10;
+    }
+
+    AP_Stats *stats = AP::stats();
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: I've made all the ones that build with const have const

File: libraries/AP_Hott_Telem/AP_Hott_Telem.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_Hott_Telem/AP_Hott_Telem.cpp]: that's not the pattern we use elsewhere, so I've kept it the same

**Pull request merged**